File: pa1/rosetta.js
// JAVASCRIPT: topological sort 

var readline = require('readline');

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false  
  // terminal=false prevents ReadLine from echoing our input back to us
});

var lines = new Array(); // we'll store all of the lines here
var edges = new Array();
var vertices = new Array();
var p_visited = new Array();// permanent mark
var t_visited = new Array();// temporary mark
rl.on('line', function(this_line) {
  // Every time there is a line available on stdin, this function will be
  // called. 
  lines.push(this_line); // add this line to what we've seen so far
}); 
rl.on('close', function() {
  // When we're finally out of input on stdin, this function will be
  // called.

	// Here we use the filter to remove duplicates in lines, there are two way 
	// to remove duplicates, but the autograder only support the next one
	// vertices = Array.from(new Set(lines));
	vertices = lines.filter(function(item, pos, self) {
		// only return element that occupy one position in the list
    	return self.indexOf(item) == pos;
	});
	vertices.sort();
	// creat edges from lines
	for (var i = 0; i < lines.length; i=i+2) {
		var edge = [lines[i],lines[i+1]];
		edges.push(edge);
		//process.stdout.write(edge+'\n');
	}
	// sort the edge list according to the second part of its elements 
	edges.sort(function(a,b) {
		var char_a = a[1][0];
		var char_b = b[1][0];
		return (char_a === char_b) ? 0 : (char_a < char_b) ? -1 : 1
	});
	vertices.forEach(function dfs(current) {
		// if permanently visited, return
		if (include(current, p_visited))
			return 0;
		// if temporarily visited, print cycle 
		if (include(current, t_visited)){
			process.stdout.write("cycle\n");
			process.exit();	
		}
		t_visited.push(current);
		edges.forEach(function(each_edge) {
			if(each_edge[0]===current){
				dfs(each_edge[1]);
			}	
		});
		p_visited.push(current);	
	});
	// print permanent mark in order
	p_visited.forEach(function(x) {
		process.stdout.write(x+'\n')	
	}); 
});

// because the autograder doesn't support the new features
// so I can only write a function to check is a list contain a specific string
function include(string, string_array) {
	var i = string_array.length;
	while(i--){	
		if (string_array[i]==string){
			return true;
		}
	
	}
	return false;
}

File: pa1/rosetta.rb
# RUBY: topological sort
require 'set'
lines = [ ]     # a list variable to hold all the lines we'll read in
working = true  # are there still more lines to read in? 
while working
  line = gets           # read a line from standard input 
  if line == nil        # nil is "nothing, it didn't work" 
    working = false     # we're done reading stuff
  else
    lines[lines.length] = line # append 'line' to the end of 'lines
  end # end of 'if'
end # end of 'while' 
$t_visited = [ ]
$p_visited = [ ]
vertices = lines.to_set
vertices = vertices.sort do |a,b| a<=>b end
edge = Struct.new(:left, :right)	#  edge struct to store left and right node
edges = []
i=0

# creat edges from standard input
until i > lines.length-2 do
	temp = edge.new(lines.at(i),lines.at(i+1))
	i = i+2
	edges.push temp
end
# sort edges alphabetically according to the right node
edges = edges.sort do |a,b| a.right <=> b.right end

def dfs(current,edges)
	if $p_visited.include?(current)
		return 0;
	end	
	if $t_visited.include?(current)
		puts "cycle\n"
		exit
	end
	$t_visited.push current
	# iterate through all the edge which start with current
	edges.each do |one_edge|
		if one_edge.left == current
			dfs(one_edge.right,edges)
		end
	end
	$p_visited.push current
	return 0;
end
vertices.each do |one_vertice| dfs(one_vertice,edges) end
$p_visited.each do |x| puts x end

